File: cisst/cisstDevices/examples/robotcomponents/scripts/WAMgc.py
from RobotComponents import *

cmnLogger.SetMask( CMN_LOG_ALLOW_ALL )
cmnLogger.SetMaskFunction( CMN_LOG_ALLOW_ALL )
cmnLogger.SetMaskDefaultLog( CMN_LOG_ALLOW_ALL )

taskManager = mtsManagerLocal.GetInstance()

rtcan = devRTSocketCAN( 'rtcan0', devCAN.RATE_1000 )

WAMqinit = array( [ 0.0, -pi/2, 0.0, pi, 0.0, 0.0, 0.0 ] )
WAM = CreateWAM( 'WAM', rtcan, WAMqinit )
WAM.Configure()
taskManager.AddComponent( WAM )

GC  = CreateGC( 'GC', '/home/sleonard/src/cisst/src/libs/etc/cisstRobot/WAM/wam7bh.rob' )

taskManager.Connect( GC.GetName(), devController.Output,
                     WAM.GetName(), devManipulator.Input )

taskManager.Connect( GC.GetName(), devController.Feedback,
                     WAM.GetName(), devManipulator.Output )

StartTasks()

print 'ENTER to enable GC'
stdin.readline()
GC.EnableCommand()

stdin.readline()

StopTasks()

File: cisst/cisstDevices/examples/robotcomponents/wam/WAMgc.cpp
#include <cisstDevices/can/devRTSocketCAN.h>

#include <cisstCommon/cmnConstants.h>

#include <cisstDevices/devKeyboard.h>
#include <cisstDevices/robotcomponents/controllers/devGravityCompensation.h>
#include <cisstDevices/robotcomponents/manipulators/Barrett/devWAM.h>

#include <cisstMultiTask/mtsTaskManager.h>

int main(){

  cmnLogger::SetMask( CMN_LOG_ALLOW_ALL );
  cmnLogger::SetMaskFunction( CMN_LOG_ALLOW_ALL );
  cmnLogger::SetMaskDefaultLog( CMN_LOG_ALLOW_ALL );



  mtsTaskManager* taskManager = mtsTaskManager::GetInstance();

  devKeyboard kb;
  kb.SetQuitKey( 'q' );
  kb.AddKeyVoidFunction('G', "ctrlenable", devController::Enable );
  kb.AddKeyVoidFunction('g', "ctrldisable", devController::Disable );
  taskManager->AddComponent( &kb );

  devRTSocketCAN can( "rtcan1", devCAN::RATE_1000 );

  vctDynamicVector<double> qinit(7, 0.0);
  qinit[1] = -cmnPI_2;
  qinit[3] =  cmnPI;

  devWAM wam( "WAM", 0.002, OSA_CPU1, &can, qinit );
  wam.Configure();
  taskManager->AddComponent( &wam );

  std::string path(CISST_SOURCE_ROOT"/libs/etc/cisstRobot/WAM/");
  vctMatrixRotation3<double> Rw0(  0.0,  0.0, -1.0, 
				   0.0,  1.0,  0.0, 
				   1.0,  0.0,  0.0 );
  vctFixedSizeVector<double,3> tw0(0.0);
  vctFrame4x4<double> Rtw0( Rw0, tw0 );
  devGravityCompensation gc( "gc", 
			     0.002,
			     devController::DISABLED,
			     OSA_CPU2,
			     path+"wam7.rob",
			     Rtw0 );
  taskManager->AddComponent( &gc );

  taskManager->Connect( kb.GetName(), "ctrlenable",
                        gc.GetName(), devController::Control );
  
  taskManager->Connect( kb.GetName(), "ctrldisable",
                        gc.GetName(), devController::Control );
  
  taskManager->Connect( gc.GetName(), devController::Output,
			wam.GetName(),devManipulator::Input );

  taskManager->Connect( gc.GetName(),   devController::Feedback,
			wam.GetName(),  devManipulator::Output );

  taskManager->CreateAll();
  taskManager->StartAll();

  pause();

  taskManager->KillAll();
  taskManager->Cleanup();


  return 0;
}
